provisioning/transport/amqp/src/ClientWebSocketTransport.cs

```diff
@@ -0,0 +1,335 @@
+        }
+
+        public override EndPoint LocalEndPoint
+        {
+            get { return this.localEndPoint; }
+        }
+
+        public override EndPoint RemoteEndPoint
+        {
+            get { return this.remoteEndPoint; }
+        }
+
+        public override bool RequiresCompleteFrames
+        {
+            get { return true; }
+        }
+
+        public override bool IsSecure
+        {
+            get { return true; }
+        }
+
+        public override void SetMonitor(ITransportMonitor usageMeter)
+        {
+            // Do Nothing
+        }
+
+        public override bool WriteAsync(TransportAsyncCallbackArgs args)
+        {
+            this.ThrowIfNotOpen();
+
+            args.Exception = null; // null out any exceptions
+
+            Task taskResult = this.WriteAsyncCore(args);
+            if (WriteTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            taskResult.ToAsyncResult(onWriteComplete, args);
+            return true;
+        }
+
+        async Task WriteAsyncCore(TransportAsyncCallbackArgs args)
+        {
+            bool succeeded = false;
+            try
+            {
+                if (args.Buffer != null)
+                {
+                    var arraySegment = new ArraySegment<byte>(args.Buffer, args.Offset, args.Count);
+                    await webSocket.SendAsync(arraySegment, WebSocketMessageType.Binary, true, writeCancellationTokenSource.Token).ConfigureAwait(false);
+                }
+                else
+                {
+                    foreach (ByteBuffer byteBuffer in args.ByteBufferList)
+                    {
+                        await webSocket.SendAsync(new ArraySegment<byte>(byteBuffer.Buffer, byteBuffer.Offset, byteBuffer.Length),
+                            WebSocketMessageType.Binary, true, writeCancellationTokenSource.Token).ConfigureAwait(false);
+                    }
+                }
+
+                succeeded = true;
+            }
+            catch (WebSocketException webSocketException)
+            {
+                throw new IOException(webSocketException.Message, webSocketException);
+            }
+#if !NETSTANDARD1_3
+            catch (HttpListenerException httpListenerException)
+            {
+                throw new IOException(httpListenerException.Message, httpListenerException);
+            }
+#endif
+            catch (TaskCanceledException taskCanceledException)
+            {
+                throw new TimeoutException(taskCanceledException.Message, taskCanceledException);
+            }
+            finally
+            {
+                if (!succeeded)
+                {
+                    this.Abort();
+                }
+            }
+        }
+
+        public override bool ReadAsync(TransportAsyncCallbackArgs args)
+        {
+            this.ThrowIfNotOpen();
+
+            Utils.ValidateBufferBounds(args.Buffer, args.Offset, args.Count);
```

Comment:
This is an internal function - we should assert instead of throwing using this validation util.
We should also try to use ArraySegment whenever possible to avoid needing to do these checks ourselves.